iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp

```diff
@@ -1289,7 +1289,17 @@ static LogicalResult createAndPropagateBufferUsedForResultTensor(
           buffer, tensorReshapeOp.reassociation());
       tensor = tensorReshapeOp.src();
       buffer = newReshapeOp.result();
-      resultTensorToBufferMap[tensor] = buffer;
+      if (resultTensorToBufferMap[tensor]) {
```

Comment:
I am not sure I fully understand the problem here. I sort of get what the issue might be, but the test case added here doesnt seem to actually test the issue as I understand it. Want to make sure I get what is being fixed.

Reply:
In the beginning, I found that it maps the result tensor of generic op to second buffer (in the test) because of the reshape op, and the first buffer is missed.

After thinking of your comment, I tested with this

```
#map0 = affine_map<(d0, d1) -> (d0, d1)>

module {
  func @store_reshape_src_and_result() {
    %c0 = constant 0 : index
    %0 = hal.interface.load.tensor @legacy_io::@arg0,
                                   offset = %c0 : tensor<2x4xf32>
    %1 = linalg.generic
           {args_in = 1 : i64, args_out = 1 : i64,
            indexing_maps = [#map0, #map0],
            iterator_types = ["parallel", "parallel"]} %0 {
    ^bb0(%arg0: f32):  // no predecessors
      %2 = tanh %arg0 : f32
      linalg.yield %2 : f32
    }: tensor<2x4xf32> -> tensor<2x4xf32>
    hal.interface.store.tensor %1, @legacy_io::@ret0, offset = %c0 : tensor<2x4xf32>
    hal.interface.store.tensor %1, @legacy_io::@ret1, offset = %c0 : tensor<2x4xf32>
    return
  }
  hal.interface @legacy_io attributes {sym_visibility = "private"} {
    hal.interface.binding @arg0, set=0, binding=0,
                                 type="StorageBuffer", access="Read"
    hal.interface.binding @ret0, set=0, binding=7,
                                 type="StorageBuffer", access="Write|Discard"
    hal.interface.binding @ret1, set=0, binding=8,
                                 type="StorageBuffer", access="Write|Discard"
  }
}
```

But it's not working, the result is

```
module {
  func @store_reshape_src_and_result() {
    %0 = iree.placeholder for "interface buffer" {binding = @legacy_io::@ret0} : memref<2x4xf32>
    %1 = iree.placeholder for "interface buffer" {binding = @legacy_io::@ret1} : memref<2x4xf32>
    %c0 = constant 0 : index
    %2 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg0} : memref<2x4xf32>
    linalg.generic {args_in = 1 : i64, args_out = 1 : i64, indexing_maps = [#map0, #map0], iterator_types = ["parallel", "parallel"]} %2, %1 {
    ^bb0(%arg0: f32, %arg1: f32):  // no predecessors
      %3 = tanh %arg0 : f32
      linalg.yield %3 : f32
    }: memref<2x4xf32>, memref<2x4xf32>
    return
  }
  hal.interface @legacy_io attributes {sym_visibility = "private"} {
    hal.interface.binding @arg0, set=0, binding=0, type="StorageBuffer", access="Read"
    hal.interface.binding @ret0, set=0, binding=7, type="StorageBuffer", access="Write|Discard"
    hal.interface.binding @ret1, set=0, binding=8, type="StorageBuffer", access="Write|Discard"
  }
}
```

Only one buffer is used to store values. I'm thinking that if there is a more general solution for this kind of issue. WIP

Reply:
THis is where the copy should happen : https://github.com/google/iree/blob/da48d03c7a87424694e833c4c36adb5a86c6d8d4/iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp#L1206 

But something is missing. I can use your example to see whats happening as well.

Reply:
@hanhanW are you looking into this. I found the issue, and I can fix it.

Reply:
This PR : https://github.com/google/iree/pull/2740 implements the fix needed here.